packages/shared/src/server/repositories/clickhouse.ts

```diff
@@ -1,34 +1,100 @@
-import { JsonNested } from "../../utils/zod";
 import { env } from "../../env";
 import { clickhouseClient } from "../clickhouse/client";
 import { logger } from "../logger";
-import { getCurrentSpan, instrumentAsync } from "../instrumentation";
+import { instrumentAsync } from "../instrumentation";
+import { S3StorageService } from "../services/S3StorageService";
+import { randomUUID } from "crypto";
 
-export const convertRecordToJsonSchema = (
-  record: Record<string, string>,
-): JsonNested | undefined => {
-  const jsonSchema: JsonNested = {};
+let s3StorageServiceClient: S3StorageService;
 
-  // if record is empty, return undefined
-  if (Object.keys(record).length === 0) {
-    return undefined;
+const getS3StorageServiceClient = (bucketName: string): S3StorageService => {
+  if (!s3StorageServiceClient) {
+    s3StorageServiceClient = new S3StorageService({
+      bucketName,
+      accessKeyId: env.LANGFUSE_S3_EVENT_UPLOAD_ACCESS_KEY_ID,
+      secretAccessKey: env.LANGFUSE_S3_EVENT_UPLOAD_SECRET_ACCESS_KEY,
+      endpoint: env.LANGFUSE_S3_EVENT_UPLOAD_ENDPOINT,
+      region: env.LANGFUSE_S3_EVENT_UPLOAD_REGION,
+      forcePathStyle: env.LANGFUSE_S3_EVENT_UPLOAD_FORCE_PATH_STYLE === "true",
+    });
   }
+  return s3StorageServiceClient;
+};
+
+export async function upsertClickhouse(opts: {
+  table: "observations" | "traces" | "scores";
+  records: Record<string, unknown>[];
+}): Promise<void> {
+  return await instrumentAsync({ name: "clickhouse-upsert" }, async (span) => {
+    // https://opentelemetry.io/docs/specs/semconv/database/database-spans/
+    span.setAttribute("ch.query.table", opts.table);
+
+    const records: Record<string, unknown>[] = opts.records.map((record) => ({
+      ...record,
+      event_ts: new Date(),
+    }));
 
-  for (const key in record) {
-    try {
-      jsonSchema[key] = JSON.parse(record[key]);
-    } catch (e) {
-      jsonSchema[key] = record[key];
+    // If event upload is enabled, we store all rows in S3 to have a backup
+    if (env.LANGFUSE_S3_EVENT_UPLOAD_ENABLED === "true") {
+      if (env.LANGFUSE_S3_EVENT_UPLOAD_BUCKET === undefined) {
+        throw new Error("S3 event store is enabled but no bucket is set");
+      }
+      const s3Client = getS3StorageServiceClient(
+        env.LANGFUSE_S3_EVENT_UPLOAD_BUCKET,
+      );
+      await Promise.all(
+        records.map((record) => {
+          s3Client.uploadJson(
+            `${env.LANGFUSE_S3_EVENT_UPLOAD_PREFIX}${record.project_id}/${record.id}/${randomUUID()}.json`,
```

Comment:
`${env.LANGFUSE_S3_EVENT_UPLOAD_PREFIX}${authCheck.scope.projectId}/${getClickhouseEntityType(type)}/${eventBodyId}/${key}.json`,

sollten wir den link genau gleich machen wie in der ingestion pipeline?